### ISSUE TITLE
Create a new coding agent like CodeAct that uses DSPy program optimization for improving the prompts.

### ISSUE THREAD

**Issue by jimwhite** [jimwhite/OpenHands#1]
Create a new coding agent like CodeAct that uses DSPy program optimization for improving the prompts.

**What problem or use case are you trying to solve?**
Create a coding agent that improves itself based on benchmarks.  In addition to (potentially) getting better agent performance this will also enable it to adapt optimally when using different LLMs (because the best prompts for each use case tends to be LLM-dependent).

**Do you have thoughts on the technical implementation?**
Basically clone CodeAct (openhands/agenthub/codeact_agent) and reimplement using DSPy (https://github.com/stanfordnlp/dspy).  The key design aspect of the agent is using DSPy Modules (https://dspy.ai/building-blocks/3-modules/).  There is an example of a coding agent using DSPy here: https://github.com/stanfordnlp/dspy/tree/main/examples/coding
